### DISCUSSION TITLE
How to create many items with unique field constraint

### DISCUSSION THREAD

**Discussion by techboy-coder** [prisma/prisma#7478] [Q&A] [answered]
Hello everybody! I'm new here. This is my first question.

I enjoy working with Prisma. It's a great dev experience. I just got a problem and haven't found any solution to this online.

---
I have got this Prisma schema.
```prisma
model Word {
  id        String   @id @default(uuid())
  text      String   @unique
  cards     Card[]
  createdAt DateTime @default(now())

  @@index([text])
}

model Card {
  id              String           @id @default(uuid())
  sides           Word[] //@unique <-- doesn't work
  card_collection CardCollection[]
  createdAt       DateTime         @default(now())

  @@index([id])
}

model CardCollection {
  id        String   @id @default(uuid())
  cards     Card[]
  createdAt DateTime @default(now())
  updatedAt DateTime @updatedAt

  @@index([id])
}
```
This schema has a unique constraint on words. Knowing, that I cannot create two same words, how can I create a/many card/s, including its words.

```typescript
    // Works if you run it once
    await prisma.card.create({
      data: {
        sides: {
          create: [{ text: "dog" }, { text: "Hund" }, { text: "perra" }]
        }
      }
    })

    // Fails if you run it again --> should only add "chien" in words and create the card
    await prisma.card.create({
      data: {
        sides: {
          create: [{ text: "dog" }, { text: "Hund" }, {text: "chien"}]
        }
      }
    })
```

Also, if I wanted to create multiple cards, while keeping a unique constraint on words and possibly words field on card, how would I go about doing so?

```typescript
// This doesn't work
await prisma.cardCollection.create({
      data: {
        cards: {
          connectOrCreate: [
            { sides: [{ text: "dog" }, { text: "Hund" }, { text: "perra" }] },
            { sides: [{ text: "cat" }, { text: "Katze" }] }
          ]
        }
      }
    })
```

I have looked in the docs. [(link)](https://www.prisma.io/docs/concepts/components/prisma-schema/relations#relations-in-the-prisma-client). 
```typescript
//From docs
const userAndPosts = await prisma.user.create({
  data: {
    posts: {
      create: [
        { title: 'Prisma Day 2020' }, // Populates authorId with user's id
        { title: 'How to write a Prisma schema' }, // Populates authorId with user's id
      ],
    },
  },
}
```
*This example (docs) only works if "title" doesn't have a unique constraint.*

If any, what's the easiest way to achieve this? If you need more info let me know. Thanks in advance.

**Comment by ryands17** [comment, marked as answer]
@techboy-coder 👋 

`@unique` doesn't work on the many side of relations, it only works on a single field like you have set the `text` in the `Word` model.

As for either creating or connecting, your syntax was incorrect. It should be like this:

```ts
await prisma.card.create({
    data: {
      sides: {
        connectOrCreate: [
          { create: { text: 'dog' }, where: { text: 'dog' } },
          { create: { text: 'Hund' }, where: { text: 'Hund' } },
          { create: { text: 'chien' }, where: { text: 'chien' } },
        ],
      },
    },
  })
```

Here's the link to the docs where you can directly search for these terms: https://www.prisma.io/docs/reference/api-reference/prisma-client-reference#connectorcreate.

**Reply by techboy-coder** [reply]
Thank you very much @ryands17! This solution works.

Is there no way to make sure that you do not have a same relationship twice. I am trying to avoid duplicates to save space. In prisma this would mean something like `Card[] @unique` or `String[] @unique`.

I was hoping prisma could use something similar to `connectOrCreate` for relationships.

**Reply by ryands17** [reply]
`connectOrCreate` is for relationships though. You can use it so that the relationship if not present will be created. There shouldn't be any duplicates using this method.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
